core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/jupyter/JupyterHtmlRenderer.kt

```diff
@@ -1,26 +1,46 @@
 package org.jetbrains.kotlinx.dataframe.jupyter
 
-import org.jetbrains.kotlinx.dataframe.DataFrame
-import org.jetbrains.kotlinx.dataframe.io.DisplayConfiguration
+import com.beust.klaxon.json
+import org.jetbrains.kotlinx.dataframe.api.rows
+import org.jetbrains.kotlinx.dataframe.api.toDataFrame
+import org.jetbrains.kotlinx.dataframe.io.*
 import org.jetbrains.kotlinx.dataframe.io.initHtml
-import org.jetbrains.kotlinx.dataframe.io.toHTML
+import org.jetbrains.kotlinx.dataframe.nrow
+import org.jetbrains.kotlinx.dataframe.size
+import org.jetbrains.kotlinx.jupyter.api.*
 import org.jetbrains.kotlinx.jupyter.api.libraries.JupyterIntegration
-import org.jetbrains.kotlinx.jupyter.api.renderHtmlAsIFrameIfNeeded
 
 internal class JupyterHtmlRenderer(
     val display: DisplayConfiguration,
     val builder: JupyterIntegration.Builder,
 )
 
 internal inline fun <reified T : Any> JupyterHtmlRenderer.render(
-    crossinline getDf: (T) -> DataFrame<*>,
     noinline getFooter: (T) -> String,
-    crossinline modifyConfig: T.(DisplayConfiguration) -> DisplayConfiguration = { it }
+    crossinline modifyConfig: T.(DisplayConfiguration) -> DisplayConfiguration = { it },
+    applyRowsLimit: Boolean = true
 ) = builder.renderWithHost<T> { host, value ->
     val contextRenderer = JupyterCellRenderer(this.notebook, host)
     val reifiedDisplayConfiguration = value.modifyConfig(display)
     val footer = getFooter(value)
-    val html = getDf(value).toHTML(
+
+    val df = convertToDataFrame(value)
+
+    val limit = if (applyRowsLimit) {
+        reifiedDisplayConfiguration.rowsLimit ?: df.nrow
+    } else {
+        df.nrow
+    }
+
+    val jsonEncodedDf = json {
+        obj(
+            "nrow" to df.size.nrow,
+            "ncol" to df.size.ncol,
+            "columns" to df.columnNames(),
+            "kotlin_dataframe" to encodeFrame(df.rows().take(limit).toDataFrame())
+        )
+    }.toJsonString()
```

Comment:
Don't serialize it here, there is no need in it

Reply:
There are a few issues:

- in the data frame there, klaxon json is used everywhere, so it needs to be converted to kotlinx.serialization to pass to kernel side functions without serialization. It certainly can be done, but I don't want to mess up with some edge cases)
- on the UI side, there are a few existing interfaces where string serialized data is expected, so it would require the serializing part of the response immediately after deserializing

Reply:
Ok, no problem then. If it won't reduce the number of changes, it doesn't worth it